examples_tests/01.HelloWorld/main.cpp

```diff
@@ -21,39 +28,282 @@ inline void debugCallback(nbl::video::E_DEBUG_MESSAGE_SEVERITY severity, nbl::vi
+	QueueFamilyIndicesArrayType queueFamilyIndices;
+
+	for (size_t i = 0ull; i < gpus.size(); ++i)
+	{
+		// Todo(achal): Hacks, get rid
+		gpu = core::smart_refctd_ptr_static_cast<video::CVulkanPhysicalDevice>(*(gpus.begin() + i));
+		auto vk_surface = core::smart_refctd_ptr_static_cast<video::ISurfaceVK>(surface);
+
+		bool isGPUSuitable = false;
+
+		// Check if the physical device has queue families which support both graphics and presenting, not necessarily overlapping
+		{
+			const auto& queueFamilyProperties = gpu->getQueueFamilyProperties();
+
+			for (uint32_t familyIndex = 0u; familyIndex < queueFamilyProperties.size(); ++familyIndex)
+			{
+				const auto& familyProperty = queueFamilyProperties.begin() + familyIndex;
+				if (familyProperty->queueFlags & video::IPhysicalDevice::E_QUEUE_FLAGS::EQF_GRAPHICS_BIT)
+					graphicsFamilyIndex = familyIndex;
+
+				if (surface->isSupported(gpu.get(), familyIndex))
+					presentFamilyIndex = familyIndex;
+
+				if ((graphicsFamilyIndex != ~0u) && (presentFamilyIndex != ~0u))
+				{
+					isGPUSuitable = true;
+					break;
+				}
+			}
+		}
+
+		// Check if this physical device supports the swapchain extension
+		// Todo(achal): Eventually move this to somewhere inside the engine
+		{
+			// Todo(achal): Get this from the user
+			const uint32_t requiredDeviceExtensionCount = 1u;
+			const char* requiredDeviceExtensionNames[] = { VK_KHR_SWAPCHAIN_EXTENSION_NAME };
 
+			uint32_t availableExtensionCount;
+			vkEnumerateDeviceExtensionProperties(gpu->getInternalObject(), NULL, &availableExtensionCount, NULL);
+			std::vector<VkExtensionProperties> availableExtensions(availableExtensionCount);
+			vkEnumerateDeviceExtensionProperties(gpu->getInternalObject(), NULL, &availableExtensionCount, availableExtensions.data());
+
+			bool requiredDeviceExtensionsAvailable = false;
+			for (uint32_t i = 0u; i < availableExtensionCount; ++i)
+			{
+				if (strcmp(availableExtensions[i].extensionName, requiredDeviceExtensionNames[0]) == 0)
+				{
+					requiredDeviceExtensionsAvailable = true;
+					break;
+				}
+			}
+
+			if (!requiredDeviceExtensionsAvailable)
+				isGPUSuitable = false;
+		}
+
+		// Check if the surface is adequate
+		{
+			std::vector<video::ISurface::SFormat> surfaceFormats =
+				gpu->getAvailableFormatsForSurface(surface.get());
+
+			std::vector<video::ISurface::E_PRESENT_MODE> presentModes =
+				gpu->getAvailablePresentModesForSurface(surface.get());
+
+			// Todo(achal): Probably should make a ISurface::SCapabilities
+			// struct for this as a wrapper for VkSurfaceCapabilitiesKHR
+			// nbl::video::ISurface::SCapabilities surfaceCapabilities = ;
+			VkSurfaceCapabilitiesKHR surfaceCapabilities;
+			vkGetPhysicalDeviceSurfaceCapabilitiesKHR(gpu->getInternalObject(),
+				vk_surface->m_surface, &surfaceCapabilities);
+
+			if ((surfaceCapabilities.maxImageCount != 0) && (SC_IMG_COUNT > surfaceCapabilities.maxImageCount)

```

Comment:
if `maxImageCount==0` but surfaceFormats and presentModes are not empty you'll never mark the GPU as unsuitable !?

Reply:
yeah, kinda janky but `maxImageCount==0` means no limit on number of swapchain available swapchain images

resolved?

Reply:
> yeah, kinda janky but maxImageCount==0 means no limit on number of available swapchain images

Well then `maxImageCount` should be very large (`~0u`) ?

Reply:
we don't store `maxImageCount` anywhere except the internal surface caps struct which get its values directly from `VkSurfaceCapabilitiesKHR `

> maxImageCount is the maximum number of images the specified device supports for a swapchain created for the surface, and will be either 0, or greater than or equal to minImageCount. A value of 0 means that there is no limit on the number of images, though there may be limits related to the total amount of memory used by presentable images.

Reply:
Resolved